### PR TITLE
Implement a loop peeling pass.

### PR DESCRIPTION
**Pull request by schweitzpgi**
This pass only implements a very resricted form of loop peeling. Specifically, a do-while loop is peeled and converted into a copy of the body of the do and a succeeding while loop.

The purpose of this transformation is to allow do-while loops to be converted to while loops. Once unrolling of general while loops is enabled, the compiler will be able to fully unroll a wider range of styles of loop statements.

### CODE REVIEW

**Review by bmhowe23 — COMMENTED** [review]
Will this work if there are `alloca`'s in the loop body? It looks like the current tests don't have any `cc.alloc`'s or `quake.alloca`'s in the loop body. I could be wrong, but it seems like special handling would be needed to handle that.
(Maybe it doesn't matter because there would also be corresponding `dealloc`'s.)

**Comment by schweitzpgi** [comment]
> Will this work if there are `alloca`'s in the loop body? It looks like the current tests don't have any `cc.alloc`'s or `quake.alloca`'s in the loop body. I could be wrong, but it seems like special handling would be needed to handle that. (Maybe it doesn't matter because there would also be corresponding `dealloc`'s.)

No special handling is needed.

A `quake.alloca` must be scoped or have a mirroring `quake.dealloc`. Specifically, it must appear within the body of a function or a `cc.scope`. Passes like add-dealloc and unwind-lowering can relax this requirement by introducing explicit `quake.dealloc` ops.

For the case in hand, any qubits allocated within the body of a do-while will be scoped or mirrored. So just like unrolling, each qubit's allocation and deallocation will happen exactly once per iteration.

File: lib/Optimizer/Transforms/LoopPeeling.cpp [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    for (Block *b = firstBlock; b != newLoopBlock; b = b->getNextNode())
+      for (auto &op : *b) {
+        if (auto contOp = dyn_cast<cudaq::cc::ContinueOp>(op)) {
+          rewriteBranch(contOp, newLoopBlock);
+          break;
+        } else if (auto brkOp = dyn_cast<cudaq::cc::BreakOp>(op)) {
+          rewriteBranch(brkOp, afterBlock);
+          break;
+        }
+      }
```
Comment (bmhowe23) [inline lib/Optimizer/Transforms/LoopPeeling.cpp:96]: Wouldn't calling `rewriter.eraseOp(op)` invalidate the iterators being used in line 88?

File: include/cudaq/Optimizer/Transforms/Passes.td [hunk 2]
```diff
@@ -182,6 +182,31 @@ def LoopNormalize : Pass<"cc-loop-normalize"> {
   ];
 }
 
+def LoopPeeling : Pass<"cc-loop-peeling"> {
```
Comment (bmhowe23) [inline include/cudaq/Optimizer/Transforms/Passes.td:185]: For my own benefit, is this a per-function pass, or a module-level pass when it isn't specified?

**Review by bmhowe23 — APPROVED** [review]

File: lib/Optimizer/Transforms/LoopPeeling.cpp (hunk 1 above)
Comment (schweitzpgi) [reply lib/Optimizer/Transforms/LoopPeeling.cpp:96]: Yep. That's why there is a `break`.

File: include/cudaq/Optimizer/Transforms/Passes.td (hunk 2 above)
Comment (schweitzpgi) [reply include/cudaq/Optimizer/Transforms/Passes.td:185]: It defaults to module from the command-line.

I think the idea is that the pass manager could schedule it either way when you leave it as a generic `Operation*`.

**Pull request merged**